### DISCUSSION TITLE
ROCm packages in [community]

### DISCUSSION THREAD

**Discussion by tpkessler** [rocm-arch/rocm-arch#937] [Announcements]
# HIP in [community]
I'm excited to tell you that today, initial support for ROCm in the official Arch Linux repositories has been released!

## What's included?
* Everything you need to render in blender with the new HIP backend;
* OpenCL runtime for darktable;
* system information with `rocminfo` and `rocm-smi`.

## Where do I find it?
* Enable the [testing repositories](https://wiki.archlinux.org/title/Official_repositories#Testing_repositories), including `[community-testing]`.
> WARNING: This may break your system!

```
pacman -Syyu && pacman -S hip-runtime-amd
```
* Add your login user to the `video` and `render` groups.
* [Optional] Add `/opt/rocm/bin` to your path.

## How do I check if ROCm is correctly installed?
* `rocminfo` collects information on your AMD GPU.
* To test HIP you can render a (simple) scene in blender or you can compile and run the following simple test
```
/opt/rocm/bin/hipcc test.cpp -o test && ./test
```
Source code
```C++
#include <iostream>
#include <cmath>
#include <vector>
#include <hip/hip_runtime.h>

__global__
void saxpy(int n, float a, const float *x, float *y)
{
    int i = hipThreadIdx_x + hipBlockDim_x * hipBlockIdx_x;
    if(i < n){
        y[i] = a * x[i] + y[i];
    }
}

__global__
void sset(int n, float a, float *x)
{
    int i = hipThreadIdx_x + hipBlockDim_x * hipBlockIdx_x;
    if(i < n){
        x[i] = a;
    }
}

int main()
{
    hipDeviceProp_t prop;
    hipGetDeviceProperties(&prop, 0);
    std::cout << "Agent " << prop.name << "\n";
    std::cout << "System version " << prop.major
        << "." << prop.minor << "\n";

    int n = 1024;
    float *x;
    float *y;
    hipMalloc((void**)&x, sizeof *x * n);
    hipMalloc((void**)&y, sizeof *y * n);

    std::vector<float> xin(n);
    for(int i = 0; i < n; i++){
        xin[i] = -1.0 + 2.0 * i / n;
    }
    hipMemcpy(x, xin.data(), sizeof *x * n, hipMemcpyHostToDevice);

    float ac = -14.412f;
    hipLaunchKernelGGL(sset, dim3(1), dim3(n), 0, 0, n, ac, y);

    float a = 5321.124f;
    hipLaunchKernelGGL(saxpy, dim3(1), dim3(n), 0, 0, n, a, x, y);

    std::vector<float> yout(n);
    hipMemcpy(yout.data(), y, sizeof *y * n, hipMemcpyDeviceToHost);

    hipFree(x);
    hipFree(y);

    for(int i = 0; i < n; i++){
        yout[i] -= a * xin[i];
        if(std::abs(yout[i] - ac) > 0.001f){
            std::cout << "Test failed at index " << i
                << " with entry " << yout[i]
                << " (" << ac << ")\n";
            return 1;
        }
    }
    std::cout << "TESTS PASSED!" << std::endl;
}
```

Example output for a Vega 56:
```
Agent AMD Radeon RX Vega
System version 9.0
TESTS PASSED!
```

## I want to do Machine Learning in X!

The ROCm stack will be released step by step. More and more packages will be added to [community] if the basic packages are sufficiently tested. As ML libraries rely on almost all ROCm packages, this will be the last step.

## Feedback
Post your experiences with the new ROCm packages below!

**Comment by jellersby** [comment]
Relating my experience so far using these packages. Thank you for putting them together! Ultimately, my goal is to build and use rocm-arch/python-pytorch-rocm.

I'm using all available packages in community-testing, and supplementing those with the remaining rocm-arch PKGBUILDs: rocm-core, rocm-dbgapi, rocm-gdb, rocm-debug-agent, rocprofiler, and roctracer.

Additionally, I had to install openmp-extras from arch4edu because makepkg fails to build it from the rocm-arch PKGBUILD.

I'm able to build rocm-arch/python-pytorch-rocm successfully! However, when loading the torch module and calling torch.cuda.is_available(), I get the following:

```
>>> import torch
>>> torch.cuda.is_available()
/usr/include/c++/12.2.0/bits/stl_vector.h:1123: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = const char*; _Alloc = std::allocator<const char*>; reference = const char*&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
zsh: IOT instruction (core dumped)  python
```

Here's a truncated backtrace from gdb:
```
#4  0x00007fffefad2112 in std::__glibcxx_assert_fail(char const*, int, char const*, char const*) (file=<optimized out>, line=<optimized out>, function=<optimized out>, condition=<optimized out>)
    at /usr/src/debug/gcc/libstdc++-v3/src/c++11/debug.cc:60
#5  0x00007fff7766ca95 in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#6  0x00007fff7766cf08 in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#7  0x00007fff7766d09c in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#8  0x00007fff7764d47f in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#9  0x00007fffa80b0b6b in  () at /opt/rocm/hip/lib/libamdhip64.so.5
#10 0x00007fffa7eaa108 in  () at /opt/rocm/hip/lib/libamdhip64.so.5
#11 0x00007fffa7ecafad in hipGetDeviceCount () at /opt/rocm/hip/lib/libamdhip64.so.5
#12 0x00007fffef9e8057 in c10::hip::device_count() () at /usr/lib/python3.10/site-packages/torch/lib/libc10_hip.so
#13 0x00007fffef3e88b2 in THCPModule_getDeviceCount_wrap(_object*, _object*) () at /usr/lib/python3.10/site-packages/torch/lib/libtorch_python.so
```

libhsa-runtime64.so seems to come from the `hip-runtime-amd` package. Could it have been built with some kind of debug assertions enabled?

Related: https://github.com/arch4edu/arch4edu/issues/203

**Reply by tpkessler** [reply]
Thank you for your feedback! Navi 3x isn't supported by the ROCm stack currently (but with the upcoming 5.5 release). I've recently added debug packages to HSA. Could you please perform an update (no need to recompile pytorch) and post the backtrace again? openmp-extras shouldn't be necessary for pytorch. I'm currently not able to build it (as of ROCm 5.4.2), which is also the main reason why it's still not added to the official repos.

**Reply by jellersby** [reply]
Hi @tpkessler, I installed the latest packages but the backtrace is the same:
```
#4  0x00007fffefad2112 in std::__glibcxx_assert_fail(char const*, int, char const*, char const*) (file=<optimized out>, line=<optimized out>, function=<optimized out>, condition=<optimized out>)
    at /usr/src/debug/gcc/libstdc++-v3/src/c++11/debug.cc:60
#5  0x00007fff7766c875 in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#6  0x00007fff7766cce8 in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#7  0x00007fff7766ce7c in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#8  0x00007fff7764d46f in  () at /opt/rocm/lib/libhsa-runtime64.so.1
#9  0x00007fffa80b0bab in  () at /opt/rocm/hip/lib/libamdhip64.so.5
#10 0x00007fffa7eaa138 in  () at /opt/rocm/hip/lib/libamdhip64.so.5
#11 0x00007fffa7ecafcd in hipGetDeviceCount () at /opt/rocm/hip/lib/libamdhip64.so.5
#12 0x00007fffef9e8057 in c10::hip::device_count() () at /usr/lib/python3.10/site-packages/torch/lib/libc10_hip.so
#13 0x00007fffef3e88b2 in THCPModule_getDeviceCount_wrap(_object*, _object*) () at /usr/lib/python3.10/site-packages/torch/lib/libtorch_python.so
```

Perhaps I failed to install the debug packages? I just did a `pacman -Sy [pkgname]` on all of the relevant ones from rocm-arch in community-testing. Updates were downloaded and installed.

**Reply by jellersby** [reply]
Hi @tpkessler, I was able to get a better backtrace from gdb. It looks like the assertion failure is happening in the hsa-rocr package in the Runtime::LoadTools() function: https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/master/src/core/runtime/runtime.cpp#L1595

```
#4  0x00007fffefad30a2 in std::__glibcxx_assert_fail(char const*, int, char const*, char const*)
    (file=file@entry=0x7fff676ea548 "/usr/include/c++/12.2.1/bits/stl_vector.h", line=line@entry=1123, function=function@entry=0x7fff676f52d0 "std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = const char*; _Alloc = std::allocator<const char*>; reference = const char*&; size_type = long unsigned i"..., condition=condition@entry=0x7fff676ea3d8 "__n < this->size()")
    at /usr/src/debug/gcc/gcc/libstdc++-v3/src/c++11/debug.cc:60
#5  0x00007fff6766c875 in std::vector<char const*, std::allocator<char const*> >::operator[](unsigned long) (this=0x7fffffffcd50, __n=0) at /usr/include/c++/12.2.1/bits/stl_vector.h:1121
#6  std::vector<char const*, std::allocator<char const*> >::operator[](unsigned long) (__n=0, this=0x7fffffffcd50) at /usr/include/c++/12.2.1/bits/stl_vector.h:1121
#7  rocr::core::Runtime::LoadTools() (this=<optimized out>) at /usr/src/debug/hsa-rocr/ROCR-Runtime-rocm-5.4.2/src/core/runtime/runtime.cpp:1595
#8  0x00007fff6766cce8 in rocr::core::Runtime::Load() (this=0x55555a01b3f0) at /usr/src/debug/hsa-rocr/ROCR-Runtime-rocm-5.4.2/src/core/runtime/runtime.cpp:1386
#9  0x00007fff6766ce7c in rocr::core::Runtime::Acquire() () at /usr/src/debug/hsa-rocr/ROCR-Runtime-rocm-5.4.2/src/core/runtime/runtime.cpp:112
#10 0x00007fff6764d46f in rocr::HSA::hsa_init() () at /usr/src/debug/hsa-rocr/ROCR-Runtime-rocm-5.4.2/src/core/runtime/hsa.cpp:206
#11 0x00007fffa72b0bab in roc::Device::init() () at /usr/src/debug/hip-runtime-amd/ROCclr-rocm-5.4.2/device/rocm/rocdevice.cpp:433
#12 amd::Device::init() () at /usr/src/debug/hip-runtime-amd/ROCclr-rocm-5.4.2/device/device.cpp:461
#13 amd::Runtime::init() () at /usr/src/debug/hip-runtime-amd/ROCclr-rocm-5.4.2/platform/runtime.cpp:75
#14 0x00007fffa70aa138 in hip::init() () at /usr/src/debug/hip-runtime-amd/hipamd-rocm-5.4.2/src/hip_context.cpp:47
#15 0x00007fffa70cafcd in hipGetDeviceCount(int*) (count=0x7fffffffda88) at /usr/src/debug/hip-runtime-amd/hipamd-rocm-5.4.2/src/hip_device_runtime.cpp:541
#16 0x00007fffeebe6cf7 in c10::hip::device_count() () at /usr/lib/python3.10/site-packages/torch/lib/libc10_hip.so
#17 0x00007fffef5b31a2 in THCPModule_getDeviceCount_wrap(_object*, _object*) () at /usr/lib/python3.10/site-packages/torch/lib/libtorch_python.so
```

**Reply by jellersby** [reply]
It looks like this is an issue upstream, and is being noticed by others: https://github.com/RadeonOpenCompute/ROCR-Runtime/issues/153

**Reply by jellersby** [reply]
@tpkessler Are the libs being built with `-D_GLIBCXX_ASSERTIONS`?

**Reply by tpkessler** [reply]
Hi @jellersby! Yes, all packages are built with glibc assertions enabled (the default on Arch Linux). I've patched the package in [community-testing]. Could you please try to run it again?

**Reply by fuyarovh** [reply]
I had the same issue and it seems to be fixed now.

**Reply by jellersby** [reply]
Thanks @tpkessler, that worked for me too. Out of curiosity, did you remove the assertions or did you patch the code?

**Reply by tpkessler** [reply]
I patched the code. Assertions are still enabled.

**Comment by tpkessler** [comment]
The full ROCm stack is now in [community], including PyTorch with ROCm support, see the official announcement on the mailing list:

https://lists.archlinux.org/archives/list/arch-dev-public@lists.archlinux.org/thread/NKV3ETKNL56NEUKT53GKRTHUANAKMHMV/

**Reply by wuxxin** [reply]
thanks for your work, pyTorch on gfx1030 is working for me,
i updated the build of torchvision for rocm  https://github.com/rocm-arch/python-torchvision-rocm 
to the corresponding pyTorch package, currently in AUR

**Reply by tpkessler** [reply]
Awesome! I'm very happy to hear that.

**Comment by sasha0552** [comment]
Can you add gfx803 to the list of PyTorch architectures? The ROCm stack can already work with gfx803, but python-pytorch-rocm cannot.
I [built](https://github.com/sasha0552/python-pytorch-rocm-gfx803/releases/tag/1.13.1-1) and tested python-pytorch-rocm with gfx803, it works.
[Just add](https://github.com/sasha0552/python-pytorch-rocm-gfx803/blob/97009ce83bb2a30832a810c02845723828db2b2b/.github/workflows/python-pytorch-rocm-gfx803.yml#L101) gfx803 to PYTORCH_ROCM_ARCH.

**Reply by tpkessler** [reply]
Hi! I've updated the package to include gfx803 and pushed it to [community-testing]. Could you please tell me if it works?

**Reply by sasha0552** [reply]
It works! Thanks

```
$ python
Python 3.10.9 (main, Dec 19 2022, 17:35:49) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.get_arch_list()
['gfx803', 'gfx900', 'gfx906', 'gfx908', 'gfx90a', 'gfx1030']
>>> torch.cuda.is_available()
True
>>> torch.cuda.get_device_name(0)
'AMD Radeon RX Graphics'
>>> d = torch.device("cuda")
>>> a = torch.rand(1, 2).to(d)
>>> print(a + 0)
tensor([[0.2274, 0.5280]], device='cuda:0')
```

**Reply by wuxxin** [reply]
i added gfx803 to the torchvision-rocm build, to mirror the arch list of pytorch-rocm

**Comment by GUUser91** [comment]
Is it possible for python-pytorch-rocm to support gfx1032?
This issue linked below says it you can use gfx1032 provided you add `export HSA_OVERRIDE_GFX_VERSION=10.3.0` to a .bashrc file
https://github.com/RadeonOpenCompute/ROCm/issues/1698

**Reply by acxz** [reply]
Moved to separate discussion for more visibility
https://github.com/orgs/rocm-arch/discussions/969

**Comment by itsagoodbrain** [comment]
@tpkessler thanks for your work getting rocm/hip into community. I didn't want to bug report as this might be expected behavior, but when I'm using pytorch from community now I get an extremely verbose output from MIOpen. The following output is repeated thousands of time every time I evaluate the model.
```
MIOpen(HIP): Info [SetStream] stream: 0, device_id: 0
MIOpen(HIP): Info [ConvolutionForward] algo = 3, workspace = 0
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1030
```
I tried exporting environmental variables to make sure the verbosity was set to debug only, but it had no effect. I'm assuming it's because those flags are for compiling not runtime. According to [these docs](https://rocmsoftwareplatform.github.io/MIOpen/doc/html/DebugAndLogging.html) the default settings would only output `Info` logs if it was a debug build.

**Reply by acxz** [reply]
Moved to a new discussion for more visibility
https://github.com/orgs/rocm-arch/discussions/970
